library/ecp.c

```diff
@@ -1945,13 +1945,13 @@ static int ecp_comb_recode_scalar( const mbedtls_ecp_group *grp,
     int ret;
     mbedtls_mpi M, mm;
 
-    mbedtls_mpi_init( &M );
-    mbedtls_mpi_init( &mm );
-
     /* N is always odd (see above), just make extra sure */
-    if( mbedtls_mpi_get_bit( &grp->N, 0 ) != 1 )
+    if( mbedtls_mpi_get_bit( &grp->N, 0 ) == 0 )
         return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
 
+    mbedtls_mpi_init( &M );
```

Comment:
I don't understand why the initialization of `M` and `mm` was moved. It makes the error case very slightly faster, but why do we care about the error case?

In general we prefer a style where objects are initialized as close as possible to the beginning of the function. That makes it easier to check that cleanup code never runs on uninitialized objects.

Reply:
> I don't understand why the initialization of `M` and `mm` was moved.

As the title says, the check was moved up because it does input parameter validation and early return in case of error.
Validation checks typically are performed before any other useful actions (in C++ it might be placed even before declarations).

> In general we prefer a style where objects are initialized as close as possible to the beginning of the function. That makes it easier to check that cleanup code never runs on uninitialized objects.

In that same general sense, cleanup code must be executed before the return, but it is skipped because as an obvious waste of time.
Similarly, in case of bad parameters. initialization would be also a waste of resources.
For example, `mbedtls_chacha20_crypt` uses a handful of `*_VALIDATE()` macros for parameter checking _before_ calling initialization function.

Reply:
Honestly I think both versions are fine in this case. There are pros and cons for each. In terms of readability, I like seeing the call to `mbedtls_mpi_init()` close to the declarations (which would mean not moving them), but OTOH as you point out since this check skips the cleanup code by symmetry it can also skip init.

Overall I think I'd still prefer not to move the init code. (I'd also like to point out that the error path is never taken, the check is really there only to make sure we remember about that assumption if we ever add short weierstrass curves with non-prime order in the future, which seems very unlikely at this point.)